### PR TITLE
RE-35 Snapshot Support

### PR DESCRIPTION
**Pull request by hughsaunders**
Please see individual commit messages, but the aims of this PR are to provide the RE infrastructure for teams to have snapshots created of their builds.

Hooks (see [RE For Projects](https://rpc-openstack.atlassian.net/wiki/spaces/RE/pages/19005457/RE+for+Projects) for more details):
* /gating/thaw/image_name: If this exists on the instance at the end of a job, (and the job uses pubcloud.runonpubcloud) then a snapshot will be created
* /gathing/thaw/run: If this exists when an instance is allocated, it will be executed. This can be used for things that need fixing when hostname/ip/etc have changed.

This PR also changes the RE_JOB_TRIGGER variable to have more granularity, the one usage of this in rpc-octavia is addressed in a PR to that repo. 

Issue: [RE-35](https://rpc-openstack.atlassian.net/browse/RE-35)

Related PRs:
* RPC-Octavia: https://github.com/rcbops/rpc-octavia/pull/21
* RPC-Openstack: https://github.com/rcbops/rpc-openstack/pull/2668

### CODE REVIEW

**Comment by hughsaunders** [comment]
recheck_unit_tests

**Comment by hughsaunders** [comment]
recheck_unit_tests

**Comment by hughsaunders** [comment]
Unit test failure is being addressed in #550

**Comment by hughsaunders** [comment]
recheck_unit_tests

**Comment by hughsaunders** [comment]
recheck_xenial_functional_test

File: rpc_jobs/snapshot.yml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+          name: BUILD_FROM_SNAPSHOT
+          description: |
+            Build instance from RPCO snapshot and run tempest.
+            Note that the IMAGE parameter should contain the
+            name of the snapshot to build from. The IMAGE
+            paramter is not passed through to the RPC AIO job,
+            which will use its own default image.
+          default: true
+      - instance_params:
+          IMAGE: "rpc_r15.0.0"
```
Comment (odyssey4me) [inline rpc_jobs/snapshot.yml]: r15.0.0 ?

**Comment by odyssey4me** [comment]
Is there a follow on card to break this out from rpc-gating to have the appropriate prep and thaw be in the target repo instead? Right now this may be fine for the immediate needs, but as soon as we need to cater for multiple branches or other projects it'll get very complex. I think it's pretty important for us to move to using in-repo scripts where possible for this process.

**Comment by odyssey4me** [comment]
And then I see https://github.com/rcbops/rpc-openstack/pull/2668 ... Note to myself - don't review big things like this when you're tired.

File: rpc_jobs/rpc_openstack.yml [hunk 2]
```diff
@@ -109,9 +109,12 @@
       - "swift"
     action:
       - deploy:
-          FLAVOR: "performance2-15"
+          FLAVOR: "7"
```
Comment (mattt416) [inline rpc_jobs/rpc_openstack.yml]: Hey @hughsaunders, why is the flavor changing here?

File: rpc_jobs/snapshot.yml (hunk 1 above)
Comment (mattt416) [reply rpc_jobs/snapshot.yml]: One thing I've not quite figured out yet is how we handle releases like newton that can be built on either Trusty or Xenial.

File: rpc_jobs/snapshot.yml [hunk 3]
```diff
@@ -0,0 +1,95 @@
+- job:
+    name: RE35-Snapshot-Test
```
Comment (mattt416) [inline rpc_jobs/snapshot.yml]: Great -- I can see this job in action!  Are you actually looking to commit this test long-term though?

File: rpc_jobs/rpc_openstack.yml (hunk 2 above)
Comment (hughsaunders) [reply rpc_jobs/rpc_openstack.yml]: Its the only flavor thats large enough and doesn't have ephemeral storage. If ephemeral storage is present, the containers are placed there, and not included in the snapshot.

File: rpc_jobs/snapshot.yml (hunk 3 above)
Comment (hughsaunders) [reply rpc_jobs/snapshot.yml]: Its been useful for testing, and may be useful for debugging and future testing.

File: rpc_jobs/snapshot.yml (hunk 1 above)
Comment (hughsaunders) [reply rpc_jobs/snapshot.yml]: This snapshot job is just for testing. Production images will be created by push triggered post merge jobs.

File: rpc_jobs/snapshot.yml (hunk 3 above)
Comment (mattt416) [reply rpc_jobs/snapshot.yml]: No objection to it going in, just wanted confirmation.

File: rpc_jobs/rpc_openstack.yml (hunk 2 above)
Comment (mattt416) [reply rpc_jobs/rpc_openstack.yml]: Building a quick rpc-openstack build [here](https://rpc.jenkins.cit.rackspace.net/job/PM_rpc-openstack-newton-xenial-swift-deploy/76/console) with a flavor 7 to see how much slower it is to complete on non-SSD.

File: rpc_jobs/snapshot.yml (hunk 1 above)
Comment (mattt416) [reply rpc_jobs/snapshot.yml]: Does this image need to be updated to reflect the distro that now gets added to the image name?

File: rpc_jobs/rpc_openstack.yml (hunk 2 above)
Comment (mattt416) [reply rpc_jobs/rpc_openstack.yml]: It only took ~ 15 mins longer, so nothing concerning.

**Comment by hughsaunders** [comment]
Update image name and help text in test job:
```diff
diff --git a/rpc_jobs/snapshot.yml b/rpc_jobs/snapshot.yml
index f8da383..f0dcad1 100644
--- a/rpc_jobs/snapshot.yml
+++ b/rpc_jobs/snapshot.yml
@@ -11,7 +11,15 @@
       - bool:
           name: CREATE_RPCO_SNAPSHOT
           default: true
-          description: Run RPCO Build and save as snapshot.
+          description: |
+            Run RPCO Build and save as snapshot. Note that the name of
+            the image that is created depends on the version of RPCO being
+            built. The build used is
+            PM_rpc-openstack-newton-xenial-swift-deploy
+            The image will be rpc_${rpc_release}_xenial.
+            rpc_release can be found in:
+            rpc-openstack/rpc-openstack/etc/openstack_deploy/group_vars/all/release.yml
+
       - bool:
           name: BUILD_FROM_SNAPSHOT
           description: |
@@ -22,7 +30,7 @@
             which will use its own default image.
           default: true
       - instance_params:
-          IMAGE: "rpc_r15.0.0"
+          IMAGE: "CHANGEME:rpc_rx.y.x_distro"
           FLAVOR: "general1-8"
           REGIONS: "DFW"
           FALLBACK_REGIONS: ""
```

**Comment by hughsaunders** [comment]
I disagree re: squashing, the commit messages give context to each change.

**Comment by hughsaunders** [comment]
recheck_xenial_functional_test

File: scripts/create_cloud_image.py [hunk 4]
```diff
@@ -70,25 +90,47 @@ def main():
...
+    print("Existing image IDs: {}".format(current_ids))
 
     conn.compute.create_server_image(
         server=args.serveruuid, name=args.imagename
     )
-
-    for image in conn.image.images():
-        if image.name == args.imagename and image.id not in current_ids:
-            saved_image = image
+    for _ in xrange(240):
```
Comment (mancdaz) [inline scripts/create_cloud_image.py]: 4 hours to find the image id after creation?

File: scripts/create_cloud_image.py [hunk 5]
```diff
@@ -70,25 +90,47 @@ def main():
...
-        sleep(60)
+        except Exception:
+            sleep(60)
+    else:
+        raise Exception("Failed to get image after creating it: {i}"
+                        .format(i=args.imagename))
+
+    print("Waiting for image:{i} to achieve active status."
+          .format(i=saved_image.name))
+    for attempt in xrange(240):
```
Comment (mancdaz) [inline scripts/create_cloud_image.py]: waiting 4 hours for an image to become active seems a bit excessive? Especially if we are waiting for this before failing the job.

File: rpc_jobs/snapshot.yml [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+                ],
+                [
+                  $class: 'StringParameterValue',
+                  name: 'FALLBACK_REGIONS',
+                  value: ""
+                ],
+                [
+                  $class: 'StringParameterValue',
+                  name: 'RPC_GATING_BRANCH',
+                  value: "RE-35"
```
Comment (odyssey4me) [inline rpc_jobs/snapshot.yml]: We'll have to change these in a follow up, but it's not crucial now.

File: scripts/create_cloud_image.py (hunk 4 above)
Comment (odyssey4me) [reply scripts/create_cloud_image.py]: Yeah, due to the size of the image and the fact that it's asynchronous, I remeber @hughsaunders mentioning that sometimes it takes a very long time.

File: scripts/create_cloud_image.py (hunk 5 above)
Comment (odyssey4me) [reply scripts/create_cloud_image.py]: Given that the it's a large image, and not cached, sometimes it takes quite a bit of time to deploy. Four hours may be a bit much, but let's see how it goes.

**Pull request merged**